op_crates/crypto/01_crypto.js

```diff
@@ -1,25 +1,211 @@
+
+  function getRandomValues(arrayBufferView) {
+    if (!ArrayBuffer.isView(arrayBufferView)) {
+      throw new TypeError(
+        "Argument 1 does not implement interface ArrayBufferView",
+      );
+    }
+    if (
+      !(
+        arrayBufferView instanceof Int8Array ||
+        arrayBufferView instanceof Uint8Array ||
+        arrayBufferView instanceof Int16Array ||
+        arrayBufferView instanceof Uint16Array ||
+        arrayBufferView instanceof Int32Array ||
+        arrayBufferView instanceof Uint32Array ||
+        arrayBufferView instanceof Uint8ClampedArray
+      )
+    ) {
+      throw new DOMException(
+        "The provided ArrayBufferView is not an integer array type",
+        "TypeMismatchError",
+      );
+    }
+    if (arrayBufferView.byteLength > 65536) {
+      throw new DOMException(
+        `The ArrayBufferView's byte length (${arrayBufferView.byteLength}) exceeds the number of bytes of entropy available via this API (65536)`,
+        "QuotaExceededError",
+      );
     }
     const ui8 = new Uint8Array(
-      typedArray.buffer,
-      typedArray.byteOffset,
-      typedArray.byteLength,
+      arrayBufferView.buffer,
+      arrayBufferView.byteOffset,
+      arrayBufferView.byteLength,
     );
     core.jsonOpSync("op_get_random_values", {}, ui8);
-    return typedArray;
+    return arrayBufferView;
   }
+
+  // Algorithm normalization, which involves storing the expected data for
+  // each pair of algorithm and crypto operation, is done on the JS side because
+  // it is convenient.
+  // Shall it stay that way, or are we better to move it on the Rust side?
+
+  // We need this method after initialization, anytime we need to normalize
+  // a provided algorithm. We store it here to prevent prototype pollution.
+  const toUpperCase = String.prototype.toUpperCase;
+
+  class RegisteredAlgorithmsContainer {
+    #nameIndex;
+    #definitions;
+
+    constructor(definitions) {
+      this.#nameIndex = Object.create(null);
+      this.#definitions = Object.create(null);
+      for (const [name, definition] of Object.entries(definitions)) {
+        this.#nameIndex[name.toUpperCase()] = name;
+        this.#definitions[name] = definition;
+      }
+    }
+
+    getDefinition(name) {
+      const upperCaseName = toUpperCase.call(name);
+      const registeredName = this.#nameIndex[upperCaseName];
+      if (registeredName === undefined) {
+        return undefined;
+      }
+      return this.#definitions[registeredName];
+    }
+
+    normalizeName(providedName) {
+      const upperCaseName = toUpperCase.call(providedName);
+      return this.#nameIndex[upperCaseName];
+    }
+  }
+
+  const supportedAlgorithms = {
+    digest: new RegisteredAlgorithmsContainer({
+      "SHA-1": {},
+      "SHA-256": {},
+      "SHA-384": {},
+      "SHA-512": {},
+    }),
+  };
+
+  const digestByteLengths = {
+    "SHA-1": 20,
+    "SHA-256": 32,
+    "SHA-384": 48,
+    "SHA-512": 64,
+  };
```

Comment:
I don't like that we need to repeat that information on the JS side. `ring` already knows about this.
I opted for copying it here and asserting that the size of the instantiated buffer matches the one expected by `ring`.
(This is why I exposed `ZeroCopyBuf.byte_length` earlier, although it is not strictly necessary)

I see two other options to avoid this:
1. Instantiate the `ArrayBuffer` on the Rust side. IMHO it is preferable, but would require significant work.
2. Expose an OP to fetch the sizes defined on the Rust side. It's not as good in terms of guarantees, but at least we don't have to hardcode the lengths like that.

Reply:
These are fixed values that will not change in the future (defined by spec), so it is fine to hard code them. Sending over the information in an extra op adds too much complexity and overhead.